libs/motor_control/esc/esc.c

```diff
@@ -0,0 +1,222 @@
+
+    if (esc->flag != ESC_INITIALIZED) {
+        printk("[ESC]: Error: Uninitialized ESC cannot be started!");
+        return -1;
+    }
+
+    if (speed > 100)
+        speed = 100;
+
+    dc_speed = (esc->dc_max - esc->dc_min) * speed / 100 + esc->dc_min;
+
+    ret = pwm_set(esc->instance.pwm_dev, esc->instance.pwm_channel,
+                  PWM_USEC(esc->period), PWM_USEC(dc_speed), 0);
+    if (ret) {
+        printk("[ESC]: Error: Selected PWM device cannot generate specified\
+        ESC signal, timer value could be overflowed due to the low prescaler\
+        value.");
+        return -1;
+    }
+
+    return 0;
+}
+
+status_t ESC_Stop(esc_t *esc) {
+    int ret;
+
+    if (esc->flag != ESC_INITIALIZED) {
+        printk("[ESC]: Error: Uninitialized ESC cannot be stopped!");
+        return -1;
+    }
+
+    ret = pwm_set(esc->instance.pwm_dev, esc->instance.pwm_channel, 0, 0, 0);
+    if (ret) {
+        printk("[ESC]: Error: Selected PWM device cannot be stopped!\n");
+        return -1;
+    }
+
+    return 0;
+}
+
+status_t ESC_DeInit(esc_t *esc) {
+    if (esc->flag != ESC_INITIALIZED) {
+        printk("[ESC]: Error: Specified ESC is already deinitialized\n");
+        return -1;
+    }
+
+    ESC_Stop(esc);
+    esc->flag = ESC_UNINITIALIZED;
+
+    return 0;
+}
+
+status_t ESC_Arm(esc_t *esc) {
+    int ret;
+
+    if (esc->flag != ESC_INITIALIZED) {
+        printk("[ESC]: Error: Uninitialized ESC cannot be armed!\n");
+        return -1;
+    }
+
+    switch (esc->protocol) {
+    case ESC_PWM:
+        ret = pwm_set(esc->instance.pwm_dev, esc->instance.pwm_channel,
+                      PWM_USEC(esc->period), PWM_USEC(ARM_DC_PWM_US), 0);
+        if (ret) {
+            printk(
+                "[ESC]: Error: Selected device cannot be armed propertly!\n");
+            ESC_Stop(esc);
+            return -1;
+        }
+
+        k_msleep(ARM_DURATION);
+
+        break;
+
+    case ESC_ONESHOT_125:
+        ret =
+            pwm_set(esc->instance.pwm_dev, esc->instance.pwm_channel,
+                    PWM_USEC(esc->period), PWM_USEC(ARM_DC_ONESHOT_125_US), 0);
+        if (ret) {
+            printk(
+                "[ESC]: Error: Selected device cannot be armed propertly!\n");
+            ESC_Stop(esc);
+            return -1;
+        }
+
+        k_msleep(ARM_DURATION);
+
+        break;
+
+    case ESC_ONESHOT_42:
+        ret = pwm_set(esc->instance.pwm_dev, esc->instance.pwm_channel,
+                      PWM_USEC(esc->period), PWM_USEC(ARM_DC_ONESHOT_42_US), 0);
+        if (ret) {
+            printk(
+                "[ESC]: Error: Selected device cannot be armed propertly!\n");
+            ESC_Stop(esc);
+            return -1;
+        }
+
+        k_msleep(ARM_DURATION);
+
+        break;
+
+    case ESC_MULTISHOT:
+        ret = pwm_set(esc->instance.pwm_dev, esc->instance.pwm_channel,
+                      PWM_USEC(esc->period), PWM_USEC(ARM_DC_MULTISHOT_US), 0);
+        if (ret) {
```

Comment:
These `if (ret)` segments can be extracted out of the switch statement since all cases call the same function.